### PR TITLE
librats: add user-defined attester, verifier, crypto and log-level

### PR DESCRIPTION
**Pull request by houhuiting**
Added the function of user-defined attester, verifier, crypto and log-level, the specific usage can be found in ./cert-app -h

### CODE REVIEW

File: include/librats/conf.h [hunk 1]
```diff
@@ -35,6 +44,7 @@ typedef struct rats_conf {
 	char crypto_type[32];
 	rats_key_algo_t key_algo;
 	rats_hash_algo_t hash_algo;
+	user_defined_param user_param;
```
Comment (haosanzi) [inline include/librats/conf.h:47]: Some fields in `user_defined_param` and `rats_conf structures` are repeated. Maybe we can directly use rats_conf to save the user defined param?

File: core/rats_init.c [hunk 2]
```diff
@@ -41,10 +41,12 @@ int common_init(rats_conf_t *conf, rats_core_context_t *ctx)
 
 	ctx->config = *conf;
 
-	rats_global_log_level = rats_loglevel_getenv("RATS_GLOBAL_LOG_LEVEL");
-	if (rats_global_log_level == (rats_log_level_t)-1) {
-		RATS_FATAL("failed to get log level from env\n");
-		rats_exit();
+	if (!conf->user_param.is_log_mode) {
+		rats_global_log_level = rats_loglevel_getenv("RATS_GLOBAL_LOG_LEVEL");
```
Comment (haosanzi) [inline core/rats_init.c:45]: Please add guides on `how to set librats library loglevel` in readme.

File: include/librats/conf.h [hunk 3]
```diff
@@ -33,6 +34,8 @@ typedef struct rats_conf {
 	char attester_type[32];
 	char verifier_type[32];
 	char crypto_type[32];
+	rats_log_level_t log_level;
+	bool is_log_mode;
```
Comment (haosanzi) [inline include/librats/conf.h:38]: Is it not possible to only  have `rats_log_level_t log_level` to set log level?

File: core/rats_init.c [hunk 4]
```diff
@@ -41,10 +41,12 @@ int common_init(rats_conf_t *conf, rats_core_context_t *ctx)
 
 	ctx->config = *conf;
 
-	rats_global_log_level = rats_loglevel_getenv("RATS_GLOBAL_LOG_LEVEL");
-	if (rats_global_log_level == (rats_log_level_t)-1) {
-		RATS_FATAL("failed to get log level from env\n");
-		rats_exit();
+	if (!conf->is_log_mode) {
+		rats_global_log_level = rats_loglevel_getenv("RATS_GLOBAL_LOG_LEVEL");
+		if (rats_global_log_level == (rats_log_level_t)-1) {
```
Comment (haosanzi) [inline core/rats_init.c:46]: please use clang-format to format your code

File: README.md [hunk 5]
```diff
@@ -108,7 +108,13 @@
  | 53         | sgx\_ecdsa                 | sgx\_ecdsa\_qve            |
 
  For instance priority, the higher, the stronger. By default, Librats will select the **highest priority** instance to use.
+ 
+ ## Run cert-app
 
+ ```
+ cd /usr/share/librats/samples
+ ./cert-app
+ ```
```
Comment (KB5201314) [inline README.md:117]: What about put those in [samples/cert-app/README.md](https://github.com/inclavare-containers/librats/blob/d841d2b27f374371e031467707d4b40c75184696/samples/cert-app/README.md)?

File: samples/cert-app/main.c [hunk 6]
```diff
@@ -193,13 +221,17 @@ int main(int argc, char **argv)
...
+	char *const short_options = "dhkc:a:v:p:l:";
 	// clang-format off
     struct option long_options[] = {
         { "debug-enclave", no_argument, NULL, 'd' },
         { "help", no_argument, NULL, 'h' },
         { "no-privkey", no_argument, NULL, 'k'},
         { "add-claim", required_argument, NULL, 'c' },
+	{ "attester", required_argument, NULL, 'a' },
+        { "verifier", required_argument, NULL, 'v' },
+	{ "crypto", required_argument, NULL, 'p' },
```
Comment (KB5201314) [inline samples/cert-app/main.c:233]: Would be better to align with rats-tls here. What about using `-c` for `crypto` and `-C` for `add-claim`?

**Review by KB5201314 — CHANGES_REQUESTED** [review]

File: include/librats/conf.h (hunk 3 above)
Comment (houhuiting) [reply include/librats/conf.h]: Ok, then I delete is_log_mode, and judge whether the user has entered log_level by judging whether log_level is RATS_LOG_LEVEL_MAX. If the user does not enter, read the log_level in the environment parameters.

File: README.md (hunk 5 above)
Comment (houhuiting) [reply README.md]: OK, I've migrated the [Run cert-app] and [Specify the instance type] sections into samples/cert-app/README.md

File: samples/cert-app/main.c (hunk 6 above)
Comment (houhuiting) [reply samples/cert-app/main.c]: Modified

Comment (houhuiting) [reply samples/cert-app/main.c]: OK, -C stands for add-claim, -c stands for crypto

**Review by KB5201314 — APPROVED** [review]

**Comment by KB5201314** [comment]
@houhuiting  LGTM

File: samples/cert-app/README.md [hunk 7]
```diff
@@ -6,6 +6,48 @@ This program will first generate the certificate in the TEE, copy it outside the
...
+### Specify the instance type
+
+The options of `cert-app` are as followed:
+
+```sh
+Options：
+
+        --debug-enclave/-d            set to enable enclave debugging
+        --no-privkey/-k               set to enable key pairs generation in librats
+        --add-claim/-C key:val        add a user-defined custom claims.
```
Comment (haosanzi) [inline samples/cert-app/README.md:25]: please remove .

**Pull request merged**